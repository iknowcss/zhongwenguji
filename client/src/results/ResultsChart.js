import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Chart } from 'chart.js';
import style from './ResultsChart.module.scss';
import colors from '../style/colors.scss';

const lightTransparent = 'rgba(' + [].slice.call(colors.light.match(/^#?(..)(..)(..)$/), 1, 4).map(n => parseInt(n, 16)).join(',') + ',.5)';

function datify(curvePoints, samplePoints) {
  let stopIndex = 0;
  curvePoints.find(([,y], i) => (stopIndex = i) && y === 0);
  stopIndex = stopIndex + 2;

  return {
    datasets: [
      {
        data: Array.from(samplePoints, ([x, y]) => ({x, y})),
        backgroundColor: 'transparent',
        borderColor: colors.accent,
        showLine: false,
        borderWidth: 2,
        pointRadius: 2
      },
      {
        showLine: true,
        data: Array.from(curvePoints.slice(0, stopIndex), ([x, y]) => ({x, y})),
        backgroundColor: lightTransparent,
        borderColor: colors.primary,
        borderWidth: 2,
        pointRadius: 0
      }
    ]
  };
}

export const renderXTick = x => Math.round(x / 100) * 100;

export const renderYTick = y => y === 100 ? '100%' : y < 100 ? `${y}` : '';

export default class ResultsChart extends Component {
  static propTypes = {
    curvePoints: PropTypes.array,
    samplePoints: PropTypes.array,
  };

  static defaultProps = {
    curvePoints: [],
    samplePoints: []
  };

  constructor() {
    super();
    this.canvasRef = React.createRef();
  }

  destroyChart() {
    if (this.chart) {
      this.chart.destroy();
    }
    this.chart = null;
    this.canvasContext = null;
  }

  initChart() {
    const { curvePoints, samplePoints } = this.props;

    this.canvasContext = this.canvasRef.current.getContext('2d');
    this.chart = new Chart(this.canvasContext, {
      responsive: true,
      type: 'scatter',
      data: datify(curvePoints, samplePoints),
      options: {
        legend: { display: false },
        scales: {
          xAxes: [{
            ticks: {
              callback: renderXTick,
              maxTicksLimit: 5,
              startAtZero: true
            }
          }],
          yAxes: [{
            ticks: {
              callback: renderYTick,
              beginAtZero: true,
              max: 120
            }
          }]
        }
      }
    });
  }

  componentDidMount() {
    this.initChart();
  }

  componentWillUnmount() {
    this.destroyChart();
  }

  render() {
    return (
      <div className={style.container}>
        <canvas ref={this.canvasRef} />
      </div>
    )
  }
}
