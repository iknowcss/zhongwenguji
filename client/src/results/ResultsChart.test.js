import React from 'react';
import { mount } from 'enzyme';
import ResultsChart, { renderXTick, renderYTick } from './ResultsChart';

const chartClassConstructorMock = jest.fn();
const chartClassDestroyMock = jest.fn();
const canvasContextMock = { canvas: 'mock' };
global.__ResultsChart = {
  chartClassConstructorMock,
  chartClassDestroyMock
};

jest.mock('chart.js', () => ({
  Chart: class {
    constructor(...args) {
      global.__ResultsChart.chartClassConstructorMock(...args);
    }

    destroy = global.__ResultsChart.chartClassDestroyMock;
  }
}));

jest.mock('../style/colors.scss', () => ({
    brandColor: '#0000ff',
    lightShades: '#fefefe',
    lightAccent: '#ffaaaa'
}));

describe('ResultsChart', () => {
  beforeAll(() => {
    chartClassConstructorMock.mockClear();
    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => canvasContextMock);
  });

  it('initalises the chart', () => {
    const curvePoints = [[0, 53], [249, 53], [498, 53]];
    const samplePoints = [[124.5, 60], [373.5, 40], [622.5, 60]];
    mount(
      <ResultsChart
        curvePoints={curvePoints}
        samplePoints={samplePoints}
      />
    );
    expect(chartClassConstructorMock).toHaveBeenCalledTimes(1);
    expect(chartClassConstructorMock.mock.calls[0])
      .toMatchSnapshot();
  });

  it('destroys the chart', () => {
    const component = mount(<ResultsChart />);

    component.unmount();

    expect(chartClassDestroyMock).toHaveBeenCalledTimes(1);
  });

  it('rounds x-ticks to the nearest hundred', () => {
    expect(renderXTick(123)).toEqual(100);
    expect(renderXTick(150)).toEqual(200);
    expect(renderXTick(178)).toEqual(200);
  });

  it('displays the y-ticks', () => {
    expect(renderYTick(0)).toEqual('0');
    expect(renderYTick(50)).toEqual('50');
    expect(renderYTick(100)).toEqual('100%');
    expect(renderYTick(101)).toEqual('');
  });
});
