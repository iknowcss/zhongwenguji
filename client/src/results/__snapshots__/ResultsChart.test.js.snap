// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsChart initalises the chart 1`] = `
Array [
  Object {
    "canvas": "mock",
  },
  Object {
    "data": Object {
      "datasets": Array [
        Object {
          "backgroundColor": "transparent",
          "borderColor": "#ff0000",
          "borderWidth": 2,
          "data": Array [
            Object {
              "x": 124.5,
              "y": 60,
            },
            Object {
              "x": 373.5,
              "y": 40,
            },
            Object {
              "x": 622.5,
              "y": 60,
            },
          ],
          "pointRadius": 2,
          "showLine": false,
        },
        Object {
          "backgroundColor": "rgba(0,255,255,0.5)",
          "borderColor": "#0000ff",
          "borderWidth": 2,
          "data": Array [
            Object {
              "x": 0,
              "y": 53,
            },
            Object {
              "x": 249,
              "y": 53,
            },
            Object {
              "x": 498,
              "y": 53,
            },
          ],
          "pointRadius": 0,
          "showLine": true,
        },
      ],
    },
    "options": Object {
      "legend": Object {
        "display": false,
      },
      "scales": Object {
        "xAxes": Array [
          Object {
            "ticks": Object {
              "callback": [Function],
              "maxTicksLimit": 5,
              "startAtZero": true,
            },
          },
        ],
        "yAxes": Array [
          Object {
            "ticks": Object {
              "beginAtZero": true,
              "callback": [Function],
              "max": 120,
            },
          },
        ],
      },
    },
    "responsive": true,
    "type": "scatter",
  },
]
`;
